File: sample/src/main/java/tech/thdev/androidrecyclerview/adapter/header_footer/CustomScrollHeaderFooterAdapter.java
package tech.thdev.androidrecyclerview.adapter.header_footer;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

import tech.thdev.androidrecyclerview.adapter.header_footer.holder.CustomScrollFooterViewHolder;
import tech.thdev.androidrecyclerview.adapter.header_footer.holder.CustomScrollHeaderViewHolder;
import tech.thdev.androidrecyclerview.adapter.header_footer.holder.CustomScrollSmallImageViewHolder;
import tech.thdev.androidrecyclerview.adapter.model.AdapterHeaderFooterContract;
import tech.thdev.androidrecyclerview.data.LocalImage;
import tech.thdev.support.widget.adapter.header_footer.BaseHeaderFooterTypedefRecyclerAdapter;

/**
 * Created by Tae-hwan on 01/11/2016.
 */

public class CustomScrollHeaderFooterAdapter extends
        BaseHeaderFooterTypedefRecyclerAdapter<LocalImage, LocalImage, Objects> implements
        AdapterHeaderFooterContract.Model<LocalImage, LocalImage, Objects>, AdapterHeaderFooterContract.View {

    public CustomScrollHeaderFooterAdapter(@NotNull Context context) {
        super(context);
    }

    @Nullable
    @Override
    public RecyclerView.ViewHolder onCreateItemViewHolder(@Nullable ViewGroup parent, int viewType) {
        return new CustomScrollSmallImageViewHolder(parent, this);
    }

    @Nullable
    @Override
    public RecyclerView.ViewHolder onCreateHeaderViewHolder(@Nullable ViewGroup parent, int viewType) {
        return new CustomScrollHeaderViewHolder(parent, this);
    }

    @Nullable
    @Override
    public RecyclerView.ViewHolder onCreateFooterViewHolder(@Nullable ViewGroup parent, int viewType) {
        return new CustomScrollFooterViewHolder(parent, this);
    }

    @Override
    public void reload() {
        notifyDataSetChanged();
    }

    @Override
    public int getHeaderItemPosition() {
        return isHeader() ? 0 : -1;
    }

    @Override
    public int getFooterItemPosition() {
        return isFooter() ? getItemCount() : -1;
    }
}

File: sample/src/main/java/tech/thdev/androidrecyclerview/ui/header_footer/custom/adapter/CustomScrollHeaderFooterAdapter.kt
package tech.thdev.androidrecyclerview.ui.header_footer.custom.adapter

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import tech.thdev.androidrecyclerview.data.LocalImage
import tech.thdev.androidrecyclerview.ui.header_footer.custom.adapter.holder.CustomScrollFooterViewHolder
import tech.thdev.androidrecyclerview.ui.header_footer.custom.adapter.holder.CustomScrollHeaderViewHolder
import tech.thdev.androidrecyclerview.ui.header_footer.custom.adapter.holder.CustomScrollSmallImageViewHolder
import tech.thdev.androidrecyclerview.ui.header_footer.custom.adapter.model.AdapterHeaderFooterContract
import tech.thdev.support.widget.adapter.header_footer.BaseHeaderFooterTypedefRecyclerAdapter
import java.util.*

/**
 * Created by Tae-hwan on 01/11/2016.
 */
class CustomScrollHeaderFooterAdapter() :
    BaseHeaderFooterTypedefRecyclerAdapter<LocalImage, LocalImage, Objects>(),
    AdapterHeaderFooterContract.Model<LocalImage, LocalImage, Objects>,
    AdapterHeaderFooterContract.View {

    override fun onCreateItemViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return CustomScrollSmallImageViewHolder(
            parent,
            this
        )
    }

    override fun onCreateHeaderViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RecyclerView.ViewHolder {
        return CustomScrollHeaderViewHolder(
            parent,
            this
        )
    }

    override fun onCreateFooterViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RecyclerView.ViewHolder {
        return CustomScrollFooterViewHolder(
            parent,
            this
        )
    }

    override fun reload() {
        notifyDataSetChanged()
    }

    override val headerItemPosition: Int
        get() = if (isHeader) 0 else -1

    override val footerItemPosition: Int
        get() = if (isFooter) itemCount else -1
}